lib/block-supports/index.php

```diff
@@ -89,11 +89,11 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
 
 	// Apply new styles and classes.
 	if ( ! empty( $new_classes ) ) {
-		$block_root->setAttribute( 'class', esc_attr( implode( ' ', $new_classes ) ) );
```

Comment:
If I am understanding correctly, it is the pre-existing classes/styles that can get messed up by `esc_attr` here since they have already had this applied?

Im assuming we still need to use `esc_attr` on what is being added in this function?  `$classes_to_add` and `$styles_to_add` may still need this before they are combined with the existing classes/styles above.

Reply:
Maybe we would have to do something similar to what we had done previously above after lines 83 and 87, something like:
`$classes_to_add = explode( ' ', esc_attr( implode( ' ', $classes_to_add ) ) );`

Reply:
I don't think we need `esc_attr` _at all_, and I believe the pre-existing values and the newly added values are in the exact same situation. The getter is aware of HTML and will return the unescaped value. So when we parse the HTML string `<div data-html-escaped="&amp;"></div>` and get the value for the `data-html-escaped` attribute, it's `&`. So whether we're extracting the values from the HTML or providing them ourselves, we'll start from the same place: unescaped values.

When we set the attributes, we provide the unescaped values and let the method deal with escaping. Note that it can do this better than `esc_attr` anyways. `esc_attr` needs to be extra careful because it's designed to operate on the values going into the attribute like this: `<div data-a="<?php echo esc_attr( '&' ); ?>"></div>`. `esc_attr` doesn't know whether it's contained in single quotes, double quotes, or no quotes at all!

When setting the attribute via the method, the DOMDocument can decide how to enclose and escape the attribute value and make a more informed decision about how the value should be escaped. It's actually this behavior that causes this bug to manifest. When we pass the html entity for apostrophe `&#039;` we see `&amp;#039;` as the rendered attribute in the generated HTML. This is a correct way of representing the unescaped _value_ `&#039;`. `setAttribute` correctly escapes it and exposes the issue with double encoding.

Here's a code example that's easy to play with:

```
<?php
// Basically what WP does
function esc_attr( $string ) { return htmlspecialchars( $string, ENT_QUOTES, 'UTF-8' ); }

$problematic_chars = '\'"<>&';
$dom = new DOMDocument( '1.0', 'utf-8' );
$dom->loadHTML( '<html><body></body></html>' );
$div = $dom->createElement( 'div' );
$dom->documentElement->firstChild->appendChild( $div );
$div->setAttribute( 'data-unescaped', $problematic_chars );
$div->setAttribute( 'data-escaped', esc_attr( $problematic_chars ) );
echo $dom->saveHtml( $div );
```

Outputs:

```html
<!-- whitespace formatting mine -->
<div
  data-unescaped="'&quot;&lt;&gt;&amp;"
  data-escaped="&amp;#039;&amp;quot;&amp;lt;&amp;gt;&amp;amp;"
></div>```

Reply:
Ohhhh, I see what you are saying.  Thanks for the detailed description!

This change makes sense to me then.